### PR TITLE
Add symbolic simulation for AES-GCM functions

### PR DESCRIPTION
**Pull request by pennyannn**
### Description:

This PR adds symbolic execution for several AES-GCM functions.

1. Added two versions of symbolic execution for `gcm_init_v8`: one proves no error and one proves some non-trivial effects.
2. Added symbolic execution for `aes_hw_encrypt` which does manual loop unrolling. This one proves no error.
3. Added an attempt for `aes_gcm_enc_kernel`. We need better loop unrolling to be able to handle this function.

### Testing:

What tests have been run? Did `make all` succeed for your changes? Was
conformance testing successful on an Aarch64 machine? Yes.

### License:

By submitting this pull request, I confirm that my contribution is
made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: Proofs/AES-GCM/GCMInitV8Sym.lean [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+      -- SP is still aligned
+    ∧ CheckSPAlignment sf
+      -- PC is updated
+    ∧ read_pc sf = read_gpr 64 30#5 s0
+    -- Htable_ptr is moved to the start of the 10th element
+    ∧ Htable_addr sf = Htable_addr s0 + (9 * 16#64)
+    -- TODO: Commenting out memory related conjuncts since it seems
+    -- to make symbolic execution stuck
+    --   -- First 12 elements in Htable is correct
+    -- ∧ read_mem_bytes 192 (Htable_addr s0) sf
```
Comment (pennyannn) [inline Proofs/AES-GCM/GCMInitV8Sym.lean:67]: Simply adding this conjunct will make `sym_n` stuck.

File: Proofs/AES-GCM/GCMInitV8Sym.lean [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    : read_err sf = .None := by
+  simp (config := {ground := true}) only at h_s0_pc
+  -- ^^ Still needed, because `gcm_init_v8_program.min` is somehow
+  --    unable to be reflected
+  sym_n 152
+  done
+
+-- TODO: this function is from Tests.Common
+-- need to put it in a better place
+def revflat (x : List (BitVec n)) : BitVec (n * x.length) :=
```
Comment (pennyannn) [inline Proofs/AES-GCM/GCMInitV8Sym.lean:37]: Where do you guys suggest putting this function in?

File: Proofs/AES-GCM/AESGCMEncKernelPre.lean [hunk 3]
```diff
@@ -0,0 +1,16 @@
+/-
+Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author(s): Yan Peng
+-/
+import Tests.«AES-GCM».AESGCMEncKernelProgram
+import Tactics.Sym
+import Tactics.StepThms
+import Correctness.ArmSpec
```
Comment (shigoel) [inline Proofs/AES-GCM/AESGCMEncKernelPre.lean:9]:
```suggestion
import Tests.«AES-GCM».AESGCMEncKernelProgram
import Tactics.StepThms
```
Just these would suffice.

File: Proofs/AES-GCM/AESGCMEncKernelSym.lean [hunk 4]
```diff
@@ -0,0 +1,56 @@
+/-
+Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author(s): Yan Peng
+-/
+import Proofs.«AES-GCM».AESGCMEncKernelPre
+import Tactics.Sym
+import Tactics.StepThms
+import Correctness.ArmSpec
+import Arm.Memory.Separate
```
Comment (shigoel) [inline Proofs/AES-GCM/AESGCMEncKernelSym.lean:10]:
```suggestion
import Proofs.«AES-GCM».AESGCMEncKernelPre
import Tactics.Sym
import Arm.Memory.Separate
```
This isn't using the new VCG stuff, so let's not include `Correctness.ArmSpec`. Also, `Tactics.StepThms` are only used in `#genStepEqTheorems`, which we already have from `AESGCMEncKernelPre`.

File: Proofs/AES-GCM/AESGCMEncKernelPre.lean (hunk 3 above)
Comment (shigoel) [reply Proofs/AES-GCM/AESGCMEncKernelPre.lean:9]: Similar comments for other files.

File: Proofs/AES-GCM/AESGCMEncKernelSym.lean [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+       ⟨(key_addr s0), 1984⟩, -- 240*8 bits key schedule and 64 bits rounds
+       ⟨(Htable_addr s0), 2048⟩, -- 16*128 bits, only first 12 elements are used
+       ⟨(out_addr s0), 128⟩ ])
+    : read_err sf = .None := by
+    simp (config := {ground := true}) only at h_s0_pc
+    sym_n 138
+    case h_s1_sp_aligned =>
+      apply Aligned_BitVecAdd_64_4
+      · assumption
+      · simp (config := {ground := true})
```
Comment (shigoel) [inline Proofs/AES-GCM/AESGCMEncKernelSym.lean:45]: Let's do `simp only` here, for robustness.

File: Proofs/AES-GCM/GCMInitV8Sym.lean [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+import Specs.GCMV8
+
+namespace GCMInitV8Program
+
+abbrev H_addr (s : ArmState) : BitVec 64 := r (StateField.GPR 1#5) s
+abbrev Htable_addr (s : ArmState) : BitVec 64 := r (StateField.GPR 0#5) s
+
+set_option debug.byAsSorry true in
+set_option maxRecDepth 1000000 in
+set_option profiler true in
```
Comment (shigoel) [inline Proofs/AES-GCM/GCMInitV8Sym.lean:18]:
```suggestion
-- set_option debug.byAsSorry true in
set_option maxRecDepth 1000000 in
-- set_option profiler true in
```

File: Proofs/AES-GCM/GCMInitV8Sym.lean [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+-- TODO: this function is from Tests.Common
+-- need to put it in a better place
+def revflat (x : List (BitVec n)) : BitVec (n * x.length) :=
+  have h : n * x.reverse.length = n * x.length := by simp only [List.length_reverse]
+  BitVec.cast h $ BitVec.flatten (List.reverse x)
+
+set_option maxRecDepth 1000000 in
+set_option maxHeartbeats 2000000 in
+set_option profiler true in
```
Comment (shigoel) [inline Proofs/AES-GCM/GCMInitV8Sym.lean:43]:
```suggestion
-- set_option profiler true in
```

File: Proofs/AES-GCM/GCMInitV8Sym.lean (hunk 2 above)
Comment (shigoel) [reply Proofs/AES-GCM/GCMInitV8Sym.lean:37]: `Arm/BitVec.lean` is the "kitchen sink" file for such bitvec utils.

**Pull request merged**